handlers/lockrenewer.go

```diff
@@ -0,0 +1,64 @@
+package handlers
+
+import (
+	"context"
+	"time"
+
+	servicebus "github.com/Azure/azure-service-bus-go"
+	"github.com/Azure/go-shuttle/tracing"
+	"github.com/devigned/tab"
+)
+
+var _ servicebus.Handler = (*PeekLockRenewer)(nil)
+
+// LockRenewer abstracts the servicebus subscription client where this functionality lives
+type LockRenewer interface {
+	RenewLocks(ctx context.Context, messages ...*servicebus.Message) error
+}
+
+// PeekLockRenewer starts a background goroutine that renews the message lock at the given interval until Stop() is called
+// or until the passed in context is canceled.
+// it is a pass through handler if the renewalInterval is nil
+type PeekLockRenewer struct {
+	next            servicebus.Handler
+	lockRenewer     LockRenewer
+	renewalInterval *time.Duration
+}
+
+func (plr *PeekLockRenewer) Handle(ctx context.Context, msg *servicebus.Message) error {
+	if plr.next == nil {
+		return NextHandlerNilError
+	}
+	if plr.renewalInterval != nil {
+		go plr.startPeriodicRenewal(ctx, msg)
+	}
+	return plr.next.Handle(ctx, msg)
+}
+
+func NewPeekLockRenewer(interval *time.Duration, lockrenewer LockRenewer, next servicebus.Handler) *PeekLockRenewer {
+	return &PeekLockRenewer{
+		next:            next,
+		lockRenewer:     lockrenewer,
+		renewalInterval: interval,
+	}
+}
+
+func (plr *PeekLockRenewer) startPeriodicRenewal(ctx context.Context, message *servicebus.Message) {
+	_, span := tracing.StartSpanFromMessageAndContext(ctx, "go-shuttle.peeklock.startPeriodicRenewal", message)
+	defer span.End()
+	for alive := true; alive; {
+		select {
+		case <-time.After(*plr.renewalInterval):
+			span.Logger().Debug("Renewing message lock")
+			err := plr.lockRenewer.RenewLocks(ctx, message)
+			if err != nil {
+				// I don't think this is a problem. the context is canceled when the message processing is over.
+				// this can happen if we already entered the interval case when the message is completing.
+				span.Logger().Info("failed to renew the peek lock", tab.StringAttribute("reason", err.Error()))
```

Comment:
warn or error?

Reply:
still feel like this should cancel the context if we repeatedly can't renew but I think we already talked about that in the other peek lock pr so not worth rehashing here.

Reply:
you can't cancel the message context. you don't have the cancel func, and
it would be really weird to pass that around IMO. you still have the same final behavior. the message is abandoned.
if you don't renew, then you still leave a chance to the handler to complete before the lock expires. but most likely, you're going to fail the Complete call too. unless it's a network fluke.
One way to potentially handle this is to put the cancel func on the context as a value? (is that possible?)
still, would be bending over backwards. I'd say we'll see if that's actually needed after some mileage.

Reply:
you don't have warn with Spans. it's error or Info. :)